ktor-client/ktor-client-core/common/src/io/ktor/client/features/HttpTimeout.kt

```diff
@@ -0,0 +1,113 @@
+         * Request timeout in milliseconds.
+         */
+        var requestTimeout: Long? = null
+
+        /**
+         * Connect timeout in milliseconds.
+         */
+        var connectTimeout: Long? = null
+
+        /**
+         * Socket timeout in milliseconds.
+         */
+        var socketTimeout: Long? = null
+
+        internal fun build(): HttpTimeout = HttpTimeout(requestTimeout, connectTimeout, socketTimeout)
+    }
+
+    /**
+     * Companion object for feature installation.
+     */
+    companion object Feature : HttpClientFeature<Configuration, HttpTimeout> {
+
+        override val key: AttributeKey<HttpTimeout> = AttributeKey("Timeout")
+
+        override fun prepare(block: Configuration.() -> Unit): HttpTimeout = Configuration().apply(block).build()
+
+        @UseExperimental(InternalCoroutinesApi::class)
+        override fun install(feature: HttpTimeout, scope: HttpClient) {
+            scope.requestPipeline.intercept(HttpRequestPipeline.Before) {
+                if (!context.attributes.contains(HttpTimeoutAttributes.key)) {
+                    context.attributes.put(HttpTimeoutAttributes.key, HttpTimeoutAttributes())
+                }
+
+                context.attributes[HttpTimeoutAttributes.key].apply {
+                    connectTimeout = connectTimeout ?: feature.connectTimeout
+                    socketTimeout = socketTimeout ?: feature.socketTimeout
+                    requestTimeout = requestTimeout ?: feature.requestTimeout
+
+                    val requestTimeout = requestTimeout ?: feature.requestTimeout
+                    if (requestTimeout == null || requestTimeout == 0L) return@apply
+
+                    val executionContext = context.executionContext!!
+                    val killer = GlobalScope.launch {
+                        delay(requestTimeout)
+                        executionContext.cancel(HttpRequestTimeoutException())
+                    }
+
+                    context.executionContext!!.invokeOnCompletion {
+                        killer.cancel()
+                    }
+                }
+            }
+        }
+    }
+}
+
+/**
+ * This exception is thrown by [HttpTimeout] to indicate timeout.
+ */
+open class HttpTimeoutCancellationException(message: String? = null) : CancellationException(message)
+
+/**
+ * This exception is thrown in case request timeout exceeded.
+ */
+class HttpRequestTimeoutException() : HttpTimeoutCancellationException("Request timeout has been expired")
+
+/**
+ * This exception is thrown in case connect timeout exceeded.
+ */
+class HttpConnectTimeoutException() : HttpTimeoutCancellationException("Connect timeout has been expired")
+
+/**
+ * This exception is thrown in case socket timeout exceeded.
+ */
+class HttpSocketTimeoutException() : HttpTimeoutCancellationException("Socket timeout has been expired")
+
+/**
+ * Container for timeout attributes to be stored in [HttpRequest.attributes].
+ */
+data class HttpTimeoutAttributes(
```

Comment:
Internal API

Reply:
It couldn't be internal so far because it's a single way for a user to specify per-request timeouts.